source/paddle.js

```diff
@@ -889,6 +922,618 @@ paddle.AttributeType = {
+                break;
+            case '2':
+                attrName = 'need_clip';
+                attrType = paddle.IR.Utility.getType('a_bool');
+                break;
+            case '3':
+                attrName = 'name';
+                attrType = paddle.IR.Utility.getType('a_str');
+                break;
+            default:
+                break;
+        }
+        attrValue = attrType === paddle.IR.Utility.getType('a_bool') ? value === 1 : value;
+        return [attrName, attrType, attrValue];
+    }
+
+    getOutputAttr(idx, outputAttr) {
+        const persistable = outputAttr[0] === 1;
+        const stop_gradient = outputAttr[1] === 1;
+        const trainable = outputAttr[2] === 1;
+        const is_distributed = outputAttr[3] === 1;
+        const is_parameter = outputAttr[4] === 1;
+        const need_clip = outputAttr[5] === 1;
+        return new Map([
+            ['persistable', paddle.IR.Utility.formatOutputAttr('persistable', persistable)],
+            ['stop_gradient', paddle.IR.Utility.formatOutputAttr('stop_gradient', stop_gradient)],
+            ['trainable', paddle.IR.Utility.formatOutputAttr('trainable', trainable)],
+            ['is_distributed', paddle.IR.Utility.formatOutputAttr('is_distributed', is_distributed)],
+            ['is_parameter', paddle.IR.Utility.formatOutputAttr('is_parameter', is_parameter)],
+            ['need_clip', paddle.IR.Utility.formatOutputAttr('need_clip', need_clip)],
+            ['name', paddle.IR.Utility.formatOutputAttr('name', outputAttr[6])]
+        ]);
+    }
+};
+
+paddle.IR.Attr = class {
+
+    constructor(idx, value, opInfo) {
+        const [attrName, attrType, attrValue] = opInfo.getAttr(idx, value);
+        this.name = attrName;
+        this.type = attrType;
+        this.value = attrValue;
+    }
+};
+
+paddle.IR.Input = class {
+    constructor(input, opInfo) {
+        const [parameter, inputName] = paddle.IR.Utility.getParaName(input, opInfo.namePrefix);
+        this.arguments = [inputName];
+        this.parameter = parameter;
+    }
+};
+
+paddle.IR.Output = class {
+
+    constructor(output, opInfo, idx, outputAttr) {
+        const [parameter, outputName] = paddle.IR.Utility.getParaName(output, opInfo.namePrefix);
+        this.arguments = [outputName];
+        this.parameter = parameter;
+        this.values = new Map();
+        this.tensorType = null;
+
+        const [, tensorType] = output.TT['#'].split('.');
+        if (tensorType === 't_dtensor') {
+            const tensorType = paddle.IR.Utility.createTensorType(output);
+            const description = opInfo.getOutputAttr(idx, outputAttr);
+            this.tensorType = tensorType;
+            this.values.set(outputName, new paddle.IR.Value(outputName, tensorType, null, description));
+        } else {
+            this.values.set(outputName, new paddle.IR.Value(outputName, null, null, null));
+        }
+
+    }
+};
+
+paddle.IR.Utility = class {
```

Comment:
Done! `paddle.IR.Utility` has been removed ~ Some static functions have been moved to `paddle.Utility`, and add an instance of `paddle.IR.GlobalInfo` holding by `paddle.IR` to handle state. Thanks!

Reply:
Thank you. Many of the types like `paddle.IR.Program` or `paddle.IR.Region` are only used while loading the file. Generally, instead of having a complete object model, could all the loading and transforming of the IR file implemented in a single transformation method to track the state. That would remove the need to track global state or having global helpers.